File: app/src/main/java/com/eds/a2appstudiointerviewassignment/WeblinkListFragment.java
package com.eds.a2appstudiointerviewassignment;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.eds.a2appstudiointerviewassignment.model.WebLink;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

public class WeblinkListFragment extends Fragment {

    public static final String TAG = "WebLink_List";
    private WebLinkListViewModel viewModel;

    public static WeblinkListFragment getFragment() {
        return new WeblinkListFragment();
    };

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_weblink_list, container, false);
        view.setTag(TAG);

        viewModel = new ViewModelProvider(requireActivity()).get(WebLinkListViewModel.class);

        WebLinkListAdapter adapter = new WebLinkListAdapter(webLink -> {
            viewModel.delete(webLink);
        });

        RecyclerView recyclerView = view.findViewById(R.id.webLinkRecyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(getContext()));
        recyclerView.setAdapter(adapter);

        viewModel.getItems().observe(getViewLifecycleOwner(), webLinks -> adapter.submitList(webLinks));

        return view;
    }
}

File: app/src/main/java/com/eds/a2appstudiointerviewassignment/weblink/WebLinkListFragment.kt
package com.eds.a2appstudiointerviewassignment.weblink

import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import com.eds.a2appstudiointerviewassignment.databinding.FragmentWeblinkListBinding

class WebLinkListFragment: Fragment(){

    private val viewModel: WebLinkViewModel by activityViewModels()

    private var _binding: FragmentWeblinkListBinding? = null
    private val binding : FragmentWeblinkListBinding
        get() = _binding!!

    companion object {
        val TAG = "WebLinkListFragment"
        fun itSelf() = WebLinkListFragment()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        _binding = FragmentWeblinkListBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        val adapter = WebLinkListAdapter()
        binding.webLinkRecyclerView.adapter = adapter
        binding.webLinkRecyclerView.layoutManager = LinearLayoutManager(requireContext())

        viewModel.list.observe(viewLifecycleOwner, Observer {
            adapter.submitList(it)
            if (it.isEmpty()) {
                binding.emptyAnimationView.visibility = View.VISIBLE
                binding.textViewStatus.visibility = View.VISIBLE
            } else {
                binding.emptyAnimationView.visibility = View.GONE
                binding.textViewStatus.visibility = View.GONE
            }
            Log.v(TAG, "Updated List => $it")
        })
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}